src/components/activities-test.js

```diff
@@ -40,78 +45,212 @@ describes.realWin('ActivityPorts test', {}, env => {
-    it('should delegate open', () => {
-      const activityPorts = new ActivityPorts(win);
-      sandbox.stub(WebActivityPorts.prototype, 'open').callsFake(() => {
-        return {targetWin: null};
+    describe('default arguments', () => {
+      let expectedDefaults;
+
+      beforeEach(() => {
+        expectedDefaults = {
+          'analyticsContext': analytics.getContext().toArray(),
+          'publicationId': pageConfig.getPublicationId(),
+          'productId': pageConfig.getProductId(),
+          '_client': 'SwG $internalRuntimeVersion$',
+        };
       });
-      const opener = activityPorts.open(
-        'some_request_id',
-        '/someUrl',
-        '_top',
-        {}
-      );
-      expect(opener.targetWin).to.be.null;
-    });
 
-    it('must delegate onResult', () => {
-      const activityPorts = new ActivityPorts(win);
-      const resultHandler = portDef => {
-        return portDef.acceptResult().then(result => {
-          expect(result.data).to.deep.equal('test');
+      it('should accept no arguments', () => {
+        expect(activityPorts.addDefaultArguments()).to.deep.equal(
+          expectedDefaults
+        );
+      });
+
+      it('should fill in an empty object', () => {
+        expect(activityPorts.addDefaultArguments({})).to.deep.equal(
+          expectedDefaults
+        );
+      });
+
+      it('should add to an existing object', () => {
+        const newArgs = {
+          aVal: 1,
+        };
+        expectedDefaults['aVal'] = 1;
+        expect(activityPorts.addDefaultArguments(newArgs)).to.deep.equal(
+          expectedDefaults
+        );
+      });
+
+      it('should not override passed values', () => {
+        const newArgs = {
+          productId: 55555555,
+        };
+        expectedDefaults['productId'] = 55555555;
+        expect(activityPorts.addDefaultArguments(newArgs)).to.deep.equal(
+          expectedDefaults
+        );
+      });
+
+      it('should not add them to open', () => {
+        let passedArgs = null;
+        sandbox
+          .stub(activityPorts.activityPorts_, 'open')
+          .callsFake((_requestId, _url, _target, args) => {
+            passedArgs = args;
+          });
+        activityPorts.open('', '', '');
+        expect(passedArgs).to.be.undefined;
+      });
+
+      it('should add them to open', () => {
+        let passedArgs;
+        sandbox
+          .stub(activityPorts.activityPorts_, 'open')
+          .callsFake((_requestId, _url, _target, args) => {
+            passedArgs = args;
+          });
+        activityPorts.open('', '', '', null, {}, true);
+        expect(passedArgs).to.deep.equal(expectedDefaults);
+      });
+
+      it('should not add them to openIframe', () => {
+        // The best test I could come up with was just to ensure it passed the
+        // arguments to addDefaultArguments and used the result.
+        let receivedArgs = null;
+        sandbox.stub(activityPorts, 'addDefaultArguments').callsFake(args => {
```

Comment:
Why can't `openIframe` be mocked?

Reply:
The purpose of the test is to ensure openIframe calls addDefaultArguments.  If I mock openIframe then it won't even call it to test it.

Reply:
You're able to mock `open` in another test though.
```js
.stub(activityPorts.activityPorts_, 'open')
```

Reply:
ActivityPorts.open calls WebActivityPorts.open.  I am mocking WebActivityPorts.open here.

openIframe call new ActivityIframePort(iframe, url, args);

If I could mock new ActivityIframePort(iframe, url, args); and pull out what it passed in as args I could do the equivalent.  I'm not sure how to mock a constructor though.

Reply:
We created a private function we could test